File: terminal/src/com/jediterm/terminal/DataStreamIteratingEmulator.java
package com.jediterm.terminal;

import com.jediterm.terminal.emulator.Emulator;

import java.io.IOException;

/**
 * @author traff
 */
public abstract class DataStreamIteratingEmulator implements Emulator {
  protected final TerminalDataStream myDataStream;
  protected final Terminal myTerminal;

  private boolean myEof = false;
  private boolean isLocked = false;

  public DataStreamIteratingEmulator(TerminalDataStream dataStream, Terminal terminal) {
    myDataStream = dataStream;
    myTerminal = terminal;

    myDataStream.addOnBeforeBlockingWaitListener(() -> {
      if (isLocked) {
        myTerminal.unlock();
        isLocked = false;
      }
    });
  }

  @Override
  public boolean hasNext() {
    return !myEof;
  }

  @Override
  public void resetEof() {
    myEof = false;
  }

  @Override
  public void next() throws IOException {
    try {
      char b = myDataStream.getChar();

      if (!isLocked) {
        myTerminal.lock();
        isLocked = true;
      }
      processChar(b, myTerminal);
    }
    catch (TerminalDataStream.EOF e) {
      if (isLocked) {
        myTerminal.unlock();
        isLocked = false;
      }
      myEof = true;
    }
  }

  protected abstract void processChar(char ch, Terminal terminal) throws IOException;
}

File: jediterm-core-mpp/src/jvmMain/kotlin/com/jediterm/terminal/DataStreamIteratingEmulator.kt
package com.jediterm.terminal

import com.jediterm.terminal.emulator.Emulator
import java.io.IOException

/**
 * @author traff
 */
abstract class DataStreamIteratingEmulator(
    protected val myDataStream: TerminalDataStream,
    protected val myTerminal: Terminal?
) : Emulator {
    private var myEof = false

    override fun hasNext(): Boolean {
        return !myEof
    }

    override fun resetEof() {
        myEof = false
    }

    @Throws(IOException::class)
    override fun next() {
        try {
            val b = myDataStream.char
            processChar(b, myTerminal)
        } catch (e: TerminalDataStream.EOF) {
            myEof = true
        }
    }

    @Throws(IOException::class)
    protected abstract fun processChar(ch: Char, terminal: Terminal?)
}
